### DISCUSSION TITLE
Strategy for typing a class that is heavily modified with a metaclass

### DISCUSSION THREAD

**Discussion by dmi-dmi-dmi** [microsoft/pyright#3774] [Q&A] [answered]
My question is a bit general but I would like to use an example with ORM for the sake of clarity.

Suppose that I use [peewee orm](https://github.com/coleifer/peewee) and have defined following model in my code

```python
from peewee import SqliteDatabase, Model, CharField, DateField

db = SqliteDatabase("people.db")


class Person(Model):
    class Meta:
        database = db

    name = CharField()
    birthday = DateField()
```

I use it in my code with the following piece
```python

p = Person()
p.name = 'John Doe'
p.birthday = date(1990, 1, 1)

p.save()
```

Pyright complains about the code
```
Cannot assign member "name" for type "Person"
  Expression of type "Literal['John Doe']" cannot be assigned to member "name" of class "Person"
    "Literal['John Doe']" is incompatible with "CharField"

Cannot assign member "birthday" for type "Person"
  Expression of type "date" cannot be assigned to member "birthday" of class "Person"
    "date" is incompatible with "DateField"
```

Internally `ModelBase` metaclass during class creation [iterates over attributes](https://github.com/coleifer/peewee/blob/0f1a49202b72b18074032efa56e42d4aeef95cce/peewee.py#L6294) of the model class and [binds all `Field`s](https://github.com/coleifer/peewee/blob/0f1a49202b72b18074032efa56e42d4aeef95cce/peewee.py#L6102) instances to the current model class and saves original instances to `Meta` class. [Binding](https://github.com/coleifer/peewee/blob/0f1a49202b72b18074032efa56e42d4aeef95cce/peewee.py#L4637) in terms of Peewee means replacing field attribute with an instance of [FieldAccessor](https://github.com/coleifer/peewee/blob/0f1a49202b72b18074032efa56e42d4aeef95cce/peewee.py#L4502) which has `__get__` and `__set__` methods.

My questions are:
1. How one should type `name` and `birthday` in this case?

Doing something like
```python
class Person(Model):
    class Meta:
        database = db

    name: str = CharField() # type: ignore
    birthday: date = DateField() # type: ignore
```
brings problem in the code that does querying

```
Person.select(Person.name).where(Person.date > ...)
```

because many methods expect instances of fields to be passed to them and pyright starts complaining here about wrong types of parameters.

Mypy seems to solve this problem with a plugin. 

2. Suppose that I want to write type stubs for Peewee. Would it be OK to simple make `Field` classes to inherit from `FieldAccessor` in the stubs? If no what approach should I take?

**Comment by erictraut** [comment, marked as answer]
How about something like this?

```python
from datetime import date
from typing import Generic, TypeVar
from peewee import SqliteDatabase, Model

db = SqliteDatabase("people.db")

_T = TypeVar("_T")

# Make FieldAccessor generic
class FieldAccessor(Generic[_T]):
    def __get__(self, instance, instance_type=...) -> _T:
        ...

    def __set__(self, instance, value: _T) -> None:
        ...

# Model field classes as though they are derived from FieldAccessor
class CharField(FieldAccessor[str]):
    ...

class DateField(FieldAccessor[date]):
    ...

class Person(Model):
    class Meta:
        database = db

    name = CharField()
    birthday = DateField()

p = Person()
p.name = "John Doe"
p.birthday = date(1990, 1, 1)
p.save()

Person.select(Person.name).where(Person.birthday > date(1990, 1, 1))
```

**Reply by dmi-dmi-dmi** [reply]
Thanks a lot for prompt response

I had to modify your snippet because actual fields are not derived from FieldAccessor. Right now it looks like this

```python
from peewee import (
    SqliteDatabase,
    Model,
    CharField as CharField_,
    DateField as DateField_,
)

...

class CharField(CharField_, FieldAccessor[str]):
    ...


class DateField(DateField_, FieldAccessor[date]):
    ...
```

So I want to be sure that it is ok to type the code like this.

What I mean is that `FieldAccessor` does not belong to the inheritance tree of almost all fields in Peewee (except maybe `ForeignKeyField` and `ManyToManyField` I believe) and its behaviour is applied dynamically in metaclass. Typing fields like in the snippet above makes sort of false claims about parents of the `DateField` and `CharField`.

My goal here is not to be 100% correct with the type hints but to make them help me catch bugs. If the snippet above doesn't add some heave side effects I am happy to use it. Basically I am a Python beginner and I wanted to know what is the recommended approach to handle some really dynamic Python code.

I do understand that it might be impossible to use type hints for some Python code due its dynamic nature and features like metaclasses so answer `no, it is impossible to type hint Peewee model to be correct in most cases` is also acceptable :-)

**Comment by erictraut** [comment]
Yeah, there are times when you need to "lie" to a type checker about the implementation to accommodate some non-standard runtime behavior. My assumption is that you plan to write type stubs for the library, and my sample above was meant to show how you could write the stub(s). The type stubs can deviate from the implementation somewhat.

As you point out, there are some libraries that are not possible to describe using the standard Python static type system because they deviate too much from standard behaviors.

[PEP 681](https://peps.python.org/pep-0681/) was an attempt to describe a family of dataclass-like behaviors that are used by many ORMs. I don't know if that helps you with peewee though. It appears that its behavior is an outlier even among ORMs.

**Reply by dmi-dmi-dmi** [reply]
Thanks a lot for clarification.

As I understand pyright already supports dataclass trasnforms so I can attempt to type my models first and then look into typing the library itself.
